### ISSUE TITLE
Feature Request: Save bone order data in MMD-tools root object, and make it accessible via bone order list in interface.

### ISSUE THREAD

**Issue by Takuyax** [powroupi/blender_mmd_tools#293]
Hi again!

I would like to discuss the way MMD-tools handles the MMD bone order.

According to the manual, currently the MMD bone order is somehow stored inside the armature modifier, which seems to be incorrect. Looking at the code, it seems it actually uses the armature's pose bone indices to get the bone order.

My problem is that this design makes re-ordering the bones in Blender impossible, and it seems that editing the armature can actually break the bone order. I had to re-order bones in pmxe a few times after export myself, and a few days ago a friend had the problem. He exported the model, and the bone order was incorrect, so he re-sorted the bones in pmx editor. I helped him re-import the pmx rig to Blender, then re-parented all his quad meshes to the re-imported armature, which is not a very beginner friendly workflow. Currently I'm helping him build a pmx modeling workflow that's supposed to become 100% Blender based. The meshes should all stay quad based, and all fixes should be done directly in the master blend file.

My suggestion would be to add a bone ordering list to the interface, the same way we already have these lists for sorting meshes and materials. That way the data wouldn't break when editing the armature, and fixes to the order could be done in a very intuitive way. I also think this handling would be much more consistent with how meshes, materials and shapekeys are sorted.

**Comment by nathanvasil** [comment]
If this is being considered, you might want to consider an alternative, which is proper and automatic bone ordering on export.  Automatic sorting of bones is something that can be done much better than automatic sorting of materials (where there's really not any kind of algorithmic answer for complicated situations.)

IIRC, the main original problem with the automatic bone sorting was its placement of IK bones chains, relative to bones that depended on those bone chains, which led to evaluation of IK only after dependent bones had been calculated.

There are potential cases where bone order couldn't be automated, but they're cases where MMD behaves very, very strangely anyways-- not situations anybody would actually want.

If you want a foolproof bone order, start at root and order in terms of parentage.  Then, shift any append bones (and their children) to after the bones from which you're appending.  Then, shift any bones that depend on any IK chain bones (including their children, including what MMD calls the "target", and their appends) to after the IK bone.  "Shift to after" should not be interpreted to ever move a bone up the list, only down the list if necessary.

Then, you need to increment deform tier.  Any bones that depend on any IK chain bone need to have their deform tier (and the deform tier of their children and appends) incremented by 1.  Then, any bones that depend on any appends from bones that are children of IK chain bones or appended by an IK chain bone (and their children, and their appends) need to have their deform tier further incremented by 1.  This process should be followed from the beginning of the list of bones to the end.

That gives you more deform tiers than you strictly need-- there are circumstances where deform tier does not need to be incremented for children of IK bones, it is only very rarely that you need to perform the last step, and I've never seen the last step necessary in anybody's models but my own, which did really weird IK stuff that nearly broke MMD-- but I believe that procedure ought to be foolproof.

Of course, with models that do things in a loop-- append from their children, for example-- this won't leave you with a useful hierarchy.  But in those cases, nothing will.

Ideally, you'd have bone ordering as suggested by Takuyax, with an automatic sort button, and an option to automatically sort on export, so that you could do it automatically, while still preserving people's ability to make utterly broken models if that's what they wanted :)

**Comment by Takuyax** [comment]
Automatic sorting sounds like a great idea. But it would be best to make it an option within the MMD-tools panel I think. That way people could sort bones automatically, and make manual changes afterwards.

Some examples for weird bone setups that require manual tweaks are those "engine" bones some models have for automatically generated breathing motions, that rely on a dependency loop to work, which Blender really hates.

And the most insane bone setup I've ever seen was on this Rin model: https://www.nicovideo.jp/watch/sm26185374 Pictures say more than words, so here is what it looks like in Blender: 
![Screenshot_20200608_215448](https://user-images.githubusercontent.com/39657731/84074384-c6b1f880-a9d2-11ea-89d2-f0c49ed57fc3.png)
Essentially she has billions of IK bones, to simulate muscle tension. I haven't tried re-importing her in MMD-tools recently, but past tries all failed (which is what you see in the screenshot), because the setup seems too MMD-specific. Blender says that she has 2799 bones.

**Comment by powroupi** [comment]
Currently, `mmd_tools` already has the ability to export custom bone order, it's not in a very intuitive way though. :thinking: 

After importing a .pmx/.pmd model, the vertex group order of Mesh object would be the bone order, and the [Armature] modifier is renamed to `mmd_bone_order_override` to tell `mmd_tools` to use the vertex group order of this Mesh object as bone order on exporting. So, **the first Mesh object which has a modifier named `mmd_bone_order_override` will be used to determine the bone order on exporting if the model has multiple Mesh objects.**

If you don't want to mess up the vertex group order of current mesh objects, attach a new mesh object with zero vertex/face to a MMD model, reorder it to the top in **Meshes Sorter** panel, add [Armature] modifier and renamed it to `mmd_bone_order_override`. Then, add bone names to the vertex groups and reorder it as you want. That's the way currently `mmd_tools` supported. :smile: 

Automatic sorting could be an option for pmx exporter, but I don't have a clear algorithm in my mind yet, in some cases may cause infinite recursion. :cry:

**Comment by nagadomi** [comment]
I explained the method mentioned by powroupi on the Japanese wiki, but I don't think many people understand this method ( Especially when modelling from scratch, because the armature  created by mmd_tools is not automatically named `mmd_bone_order_override` on modifier panel ).

I think it will be easier to use if mmd_tools has the same interface as  Mesh sorter/Material sorter for bone order.  Also PMXE has a similar interface.

**Comment by nathanvasil** [comment]
@powroupi Thanks for the info, that's useful.  What about the cases where there is no vertex group for a particular bone?  (Non-deforming bones, like a motherbone for example.)

@Takuyax Thanks, that's interesting!  I had always assumed that the breathing was done via perpetual motion physics and "after physics" bones.

**Comment by Takuyax** [comment]
@powroupi Thanks for explaining. In my workflow I usually eliminate unassigned vertex groups, so this explains the cases where the bone order broke. My models are split into different mesh objects for body, hair and clothing parts, so none of them usually have the full vertex group list. Will the technique you described also work when using an empty (zero vertices) mesh for the bone order?

@nathanvasil I would be lying if I claimed to know all the ways this could be done, but I'm pretty sure I've seen a few rigs where it was only a dependency loop of bones moving each other. Didn't you make a girl who starts dancing when only a single bone is moved? Uzuki if I remember correctly? She is pretty cool.

**Comment by nathanvasil** [comment]
@takuyax I did make the model, didn't realize anybody ever looked at it lol :)  Kind of an interesting learning experience, but complexity is too high for implementation or use.  These days, I'd much rather just use Blender's tools (and possibly bake visual animations if I didn't want to render in Blender) which allow for a much higher complexity ceiling.  It's hard to try to do everything with appends and IKs....  Currently tossing around how I'm going to handle a stretch to constraint (for knee volume preservation) in native, raw MMD.  It's not impossible, but it'll take some experimentation.

**Comment by powroupi** [comment]
>Will the technique you described also work when using an empty (zero vertices) mesh for the bone order?

Yes, the only thing `mmd_tools` needs is the vertex group name list of a mesh object. `mmd_tools` will try to find the first mesh object with a modifier named `mmd_bone_order_override` in the order shown in Meshes Sorter panel, it will sort Blender's bone list by the vertex group name list of that mesh object.

>What about the cases where there is no vertex group for a particular bone?

The bone names which cannot be found in the vertex group will be ranked at the end like newly added bones, so you will need to define full name list to get correct/desired bone order.

>I think it will be easier to use if mmd_tools has the same interface as Mesh sorter/Material sorter for bone order. Also PMXE has a similar interface.

The downside is that Blender's UIList is not very useful for rearranging items, especially for large list, multi-select is your best friend. The most efficient way would be to simply use Text Editor I think.

**Comment by summer-candy** [comment]
![image](https://user-images.githubusercontent.com/54714430/84238141-215f6780-ab2d-11ea-9c9f-86f93c44e080.png)
try this

**Comment by powroupi** [comment]
Okay, I just added a Bone Order panel shows the basic information (Parent Bone, Append Bone, IK Bone, After Physics, Deformation Tier) with simple error check for easier reviewing the bone order of active MMD model. :smile:

Not sure if the error check is correct or not. I am looking for general rules. :thinking: 
 - (bone.index + bone.deformation_tier * total_bone_count) should greater than (parent.index + parent.deformation_tier * total_bone_count)
 - (bone.index + bone.deformation_tier * total_bone_count) should greater than (append.index + append.deformation_tier * total_bone_count)
 - (bone.index + bone.deformation_tier * total_bone_count) should less than (ik_bone.index + ik_bone.deformation_tier * total_bone_count)

**Comment by nathanvasil** [comment]
@powroupi Mmm, no, not in a strict mathematical sense.

A bone can be before its parent if it has a higher deform tier.  In that case, it doesn't at all matter *how* much before its parent it is.  And I'm not sure where total bone count figures in.

In pseudo code:

(bone.parentBone.id < bone.id) AND (bone.parentBone.deformtier <= bone.deformtier)

AND

(bone.appendBone.id < bone.id) OR (bone.appendBone.deformtier < bone.deformtier)

AND

(bone.IKchainBone[n].id < bone.id)

AND

(bone.IKchain.IKtargetBone.id > bone.id)

AND

(NOT bone.parentBone.isIKchained) OR (bone.parentBone.IKchain.IKtargetBone.id = bone.id) OR ( (bone.parentBone.id < bone.id) AND (bone.parentBone.deformtier < bone.deformtier))

That's off the top of my head, and it's stricter than you'd need for many models.  Hopefully it makes sense?

**Comment by powroupi** [comment]
@nathanvasil thanks for the information. :smile: 

>(bone.appendBone.id < bone.id) OR (bone.appendBone.deformtier < bone.deformtier)

Do you mean
```
{ (bone.appendBone.id < bone.id) AND (bone.appendBone.deformtier <= bone.deformtier) }
OR
{ (bone.appendBone.id > bone.id) AND (bone.appendBone.deformtier < bone.deformtier) }
```

>(bone.IKchainBone[n].id < bone.id)

Should it be the same way like appendBone?
```
{ (bone.IKchainBone[n].id < bone.id) AND (bone.IKchainBone[n].deformtier <= bone.deformtier) }
OR
{ (bone.IKchainBone[n].id > bone.id) AND (bone.IKchainBone[n].deformtier < bone.deformtier) }
```

>(bone.IKchain.IKtargetBone.id > bone.id)

The bone orders `[IKChain, IKTarget, IKBone]`, `[IKChain, IKBone, IKTarget]` are both possibly working fine with the same deform tier, I'm not sure which way is better? I thought it should be the same rule with IKChain. :confused: 

>(NOT bone.parentBone.isIKchained) OR (bone.parentBone.IKchain.IKtargetBone.id = bone.id) OR ( (bone.parentBone.id < bone.id) AND (bone.parentBone.deformtier < bone.deformtier))

I kind of understand this.
`(bone.parentBone.IKchain.IKtargetBone.id = bone.id)` Do you mean bone.parentBone's IKBone.IKchain.IKtargetBone.id = bone.id ?

**Comment by nathanvasil** [comment]
@powroupi 

> Do you mean

Yes, thanks for the catch.  (In English: a bone needs have a deform tier equal to greater than its append target and be later in the hierarchy, or else it needs to have a deform tier greater than its append target.)

> Should it be the same way like appendBone?

Sorry, another thing I missed.  Yes, the IK target needs to have a deform tier equal to or greater than the bones involved in the chain.

However, I think there might be problems when using a deform tier that's greater than the chain bones, but I can't remember exactly.  Ideally, the IK bone should have the same deform tier as the bones in the chain, and that should always be achievable in a non-looping structure.

> I'm not sure which way is better?

I don't think I understand the question, but let me rephrase what I wrote in English, because I think this is a case where English would work better :)

I'm giving a recommended order for IK.  Let's take an example: "femur"->"calf" is the IK chain for IK bone "legIK"; the target for legIK is "ankle" which is parented to calf; and there is an additional bone "thighBauble" that is parented to femur.

In that case, the safest order of evaluation is femur, calf, legIK, ankle, thighBauble, in that order.  Assuming we start at deform tier 0, then femur through ankle should have a deform tier of 0 while thighBauble should have a deform tier of 1.

This is one of those cases where it's stricter than it needs to be.  ThighBauble doesn't actually need an incremented deform tier unless thighBauble itself is part of an IK chain.  But it never hurts to increment it, and it's simpler to explain and evaluate this way.

Likewise, this particular ordering of IK (chain, IK, target) is not always necessary, but it *sometimes* is, and it shouldn't ever cause any problems.  (You'll see a different structure in many models, but that structure could safely be reordered into what I'm recommending.)

> IKBone.IKchain.IKtargetBone.id = bone.id ?

Yeah, sorry, that's one of the places where the English is better.  I was trying to get at the order I described last paragraph.  Put the chain bones before the IK bone, and put the IK bone before the IK target.

(One of the things that confuses the language here is the difference in the way Blender and PMXE translations use the word "IK target."  In Blender, "calf" would be the IK bone and "legIK" would be the IK target.  In MMD, "legIK" is the IK bone and "ankle" is the IK target.)

**Comment by powroupi** [comment]
Great, now I understand the strict rules you described, thank you. :smile: But the case of thighBauble might not be easy to detect, the calf would need to have a deform tier higher than femur which is not what we want. And what if the case that a bone is IK chained by multiple IK bones, maybe another femurIK controls the femur's angle? :thinking: 

I think the Bone Order panel is currently working fine with non-strict error check, except IKTarget which is using the order `(chain, target, IK)`. (because many models have the order of evaluation like `femur, calf, ankle, legIK`), ... maybe I'll keep it with non-strict error check as default and add an option for strict error check in the future. :smile:

**Comment by nathanvasil** [comment]
@powroupi The calf doesn't need to have a deform tier higher than thigh.  It does not depend on thighBauble.  Both calf and thighBauble are children of femur.  (And if you wanted to append calf from thighBauble, you'd have a loop, because the position of thighBauble depends on the IK, which depends on the orientation of calf.)

The safe rule is, increment the deform tier of children of IK chain bones, and of bones that append from IK chain bones (and of their dependents as well: children, children of children, bones that append from children of children.....)  If you create a structure where you backlink appends, it shouldn't be that complicated, you just follow the hierarchy and the backlinks until you run out of bones.

I haven't considered the possibility of multiple IK affecting a single bone.  I've never seen a model in MMD do that, and I'm not sure how it would be handled.  I would expect it to cause some frame lag no matter what you did (and I consider frame lag to mean, "this isn't right.").  When I've made some "IK guides" for later IKs, I've created duplicated bones so that there were no shared bones in any of the chains.

I don't expect chain, target, IK to be a problem.  I think there was probably a reason I settled on chain, IK, target in my own models, but I honestly can't remember what it was, and it was likely some really weird corner case.

**Comment by powroupi** [comment]
>The calf doesn't need to have a deform tier higher than thigh.

Yeah, I know. I mean the calf would need to have a deform tier higher than femur when we check calf with this rule.
```
(NOT bone.parentBone.isIKchained)
OR (bone.parentBone.IKchain.IKtargetBone.id = bone.id)
OR ( (bone.parentBone.id < bone.id) AND (bone.parentBone.deformtier < bone.deformtier) )
```

>I don't expect chain, target, IK to be a problem.

Yeah, I mean currently Bone Order panel will show error icon on the target of the order `chain, IK, target`. That might not be ideal.

**Comment by nathanvasil** [comment]
>Yeah, I know. I mean the calf would need to have a deform tier higher than femur when we check calf with this rule.

I'm sorry, I phrased that wrong.  I shouldn't have even tried.  I hope I'm not just screwing you up :(

What I was trying to get at was the order I recommend of chain, chain, (chain...), IK, target, child of chain.  I was trying to describe the rule for thighBauble.

But, yeah, at the very least, chain, IK, target is definitely safe.  (Works on my own models, and no PMXE error warnings.)

**Comment by rintrint** [comment]
@powroupi I want to know why MMD tools don't use the original bone_id and instead maintain a new bone_id for referencing bone morphs, rotate+, and move+ operations. If MMD tools used the original bone_id when importing models, it would be much easier to preserve bone order.
I want to modify MMD tools to use the original bone_id, but I don't understand the reason behind the current implementation, so I'm worried my efforts might be wasted.

<img width="264" alt="Image" src="https://github.com/user-attachments/assets/a0e5ee24-96fe-4fba-a1f6-5c1fe471baa0" />
